### PR TITLE
matcher: add prefix matching support for matching API

### LINKED ISSUE #20198: matcher: support prefix match map for prefix matching
*Title*: matcher: support prefix match map for prefix matching

*Description*:

https://github.com/envoyproxy/envoy/blob/9adf1704c62725b324112beec2a50fbd1fa602ed/source/common/matcher/matcher.h#L225-L226

As mentioned in #6602 and #20147, we would like to have a prefix matching mechanism for route matching. Fortunately, the unified matching API provides an interface for prefix matching, but the code has not been implemented in Envoy. It would be good if we can support the prefix matching which can potentially help improving the performance of Envoy.

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: matcher: add prefix matching support for matching API
Additional Description: Envoy has the matching API support for exact map and prefix map while the prefix map has not been implemented yet. This patch realizes a general prefix map matcher based on trie lookup table for prefix matching. The prefix map matcher is not been optimized for IP (like the trie matcher based on LC trie) or path (like the path-splitting one).
Risk Level: Low or medium
Testing: Unit and integration
Docs Changes: N/A
Release Notes: WIP
Platform Specific Features: N/A
Fixes #20198

### CODE REVIEW

**Review by snowp — CHANGES_REQUESTED** [review]
This looks great, thanks for working on this! A few comments to reduce code duplication between this and the other match tree type

File: source/common/matcher/prefix_map_matcher.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#include "envoy/matcher/matcher.h"
+
+#include "source/common/common/utility.h"
+
+namespace Envoy {
+namespace Matcher {
+
+/**
+ * Implementation of a trie match tree that provides lookup of prefix values,
+ * with one longest prefix OnMatch per result.
```
[the remark this answers is not in the document]
Comment (zhxie) [reply source/common/matcher/prefix_map_matcher.h]: Ok, changed.

File: source/common/matcher/matcher.h [hunk 2]
```diff
@@ -222,8 +223,26 @@ class MatchTreeFactory : public OnMatchFactory<DataType> {
         return multimap_matcher;
       };
     }
-    case MatcherType::MatcherTree::kPrefixMatchMap:
-      PANIC("unexpected matcher type");
+    case MatcherType::MatcherTree::kPrefixMatchMap: {
```
[the remark this answers is not in the document]
Comment (zhxie) [reply source/common/matcher/matcher.h:210]: Ok, done.

File: source/common/matcher/prefix_map_matcher.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        return {MatchState::MatchComplete, OnMatch<DataType>{result->action_cb_, nullptr}};
+      }
+    } else if (input.data_availability_ ==
+               DataInputGetResult::DataAvailability::MoreDataMightBeAvailable) {
+      // It's possible that we were attempting a lookup with a partial value, so delay matching
+      // until we know that we actually failed.
+      return {MatchState::UnableToMatch, absl::nullopt};
+    }
+
+    return {MatchState::MatchComplete, on_no_match_};
```
[the remark this answers is not in the document]
Comment (zhxie) [reply source/common/matcher/prefix_map_matcher.h]: Ok, done.

**Review by snowp — APPROVED** [review]
Nice, LGTM!

**Comment by snowp** [comment]
@yanavlasov for second set of eyes

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**